File: Week_02/G20200343030632/Leetcode_70_632.py
#
# @lc app=leetcode.cn id=70 lang=python3
#
# [70] 爬楼梯
#

# @lc code=start
class Solution:
    memory = {}
    def climbStairs(self, n: int) -> int:
        """
            递归解决:
            归纳条件: f(n) = f(n - 1) + f(n - 2)
            1. 退出条件: n <= 2 return n
            2. 加了一个缓存优化一下
        """
        if n <= 2:
            return n
        if n - 2 in Solution.memory:
            res1 = Solution.memory.get(n - 2)
        else:
            res1 = self.climbStairs(n - 2)
            Solution.memory[n - 2] = res1
        if n - 1 in Solution.memory:
            res2 = Solution.memory.get(n - 1)
        else:
            res2 = self.climbStairs(n - 1)
            Solution.memory[n - 1] = res2
        return res1 + res2
# @lc code=end

File: Week_02/G20200343030632/Leetcode_70_632.java
import java.util.Map;

class Solution {
    Map<Integer, Integer> memory = new HashMap<Integer, Integer>();
    
    public int climbStairs(int n) {
        // f(n) = f(n - 2) + f(n - 1)
        // f(3) = f(2) + f(1)
        // f(2) = 2, f(1) = 1
       if (n < 3) {
           return n;
       }
       int first = 1;
       int second = 2;
       int ret = 0;
       for (int i = 3; i <= n; i++) {
            ret = first + second;
            first = second;
            second = ret;
       }
       return second;
    }
}